### PR TITLE
CAT-2296: Improve UX for web loading screens

### PR DESCRIPTION
**Pull request by FleetAdmiral**
Made the changes as per CAT-2297 (under CAT-2296).

@robertpainsi This is the new PR, please do look into it. Thanks :)

### CODE REVIEW

**Comment by robertpainsi** [comment]
Just two little things you already did in your previous PR but missed in this one somehow.
* Add the jira ticket to the commit message. (CAT-…)
* Do not abbreviate words. At least not background to bg. It's easier to search the history later on.

Resulting in something like:
`CAT-2297 Changed background color of webview`

**Comment by FleetAdmiral** [comment]
@robertpainsi Will keep in mind from next time, slipped my mind somehow yesterday :P. Can i leave it like this, or should i make a dummy commit?

**Comment by robertpainsi** [comment]
Just reword your commit message and force push to your branch. No need for a dummy commit or new pull request :wink: 
https://help.github.com/articles/changing-a-commit-message

**Comment by FleetAdmiral** [comment]
@robertpainsi  Done :). Also, i'd like to know a bit on how the team plans to redesign the UX so that i can add a few sample code snippets in my GSoC application according to that. Where can i ask regarding the same? (I'm guessing the google groups, but wouldn't irc be more convenient if i have doubts?). Thanks 👍

**Comment by FleetAdmiral** [comment]
@robertpainsi I've committed CAT-2298 onto the same PR, do let me know if any issues persist. Now onto the last issue in CAT-2296.

**Comment by FleetAdmiral** [comment]
@robertpainsi I've committed all 3 subtasks. Do look into it :)

**Comment by robertpainsi** [comment]
Awesome! :smiley_cat: That was really fast. I'll review your changes asap

> …how the team plans to redesign the UX… google groups

Yes, please use google groups for any questions addressing gsoc17.

File: catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java [hunk 1]
```diff
@@ -64,13 +66,15 @@
...
 	private static final String FILENAME_TAG = "fname=";
+	private ProgressDialog progressCircle;
 
 	private WebView webView;
 	private boolean callMainMenu = false;
 	private String url;
 	private String callingActivity;
 	private ProgressDialog progressDialog;
 	private Intent resultIntent = new Intent();
+    private int webload_status = 0;
```
Comment (FleetAdmiral) [inline catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java]: Yea i'd initially done it with boolean, must have changed it later for some reason. I'll change it back to boolean.. Any other changes? I'll make them all together :)

**Comment by cheidenreich** [comment]
Hi @FleetAdmiral, Thanks for your PR.

I've run your code on our test server, please fix the following warnings ;)

```
WebViewActivity.java:49, UnusedImportsCheck, Priorität: Normal
Unused import - android.widget.Toast.

WebViewActivity.java:77, RegexpSinglelineJavaCheck, Priorität: Normal
Spaces arent allowed for indentation. Tabs only

WebViewActivity.java:77, IndentationCheck, Priorität: Normal
'member def modifier' have incorrect indentation level 4, expected level should be 1.

WebViewActivity.java:77, MemberNameCheck, Priorität: Normal
Member name not following naming convention - Name 'webload_status' must match pattern '^(([a-z]{2})|([x-z][A-Z]))[a-zA-Z0-9]*$'.

WebViewActivity.java:97, RegexpMultilineCheck, Priorität: Normal
Maximum of blank lines is 1

WebViewActivity.java:105, RegexpSinglelineJavaCheck, Priorität: Normal
Spaces arent allowed for indentation. Tabs only

WebViewActivity.java:105, IndentationCheck, Priorität: Normal
'if' child have incorrect indentation level 40, expected level should be one of the following: 5, 6, 7.

WebViewActivity.java:113, IndentationCheck, Priorität: Normal
'if' have incorrect indentation level 16, expected level should be one of the following: 4, 5, 6.

WebViewActivity.java:113, RegexpSinglelineJavaCheck, Priorität: Normal
Spaces arent allowed for indentation. Tabs only

WebViewActivity.java:114, RegexpSinglelineJavaCheck, Priorität: Normal
Spaces arent allowed for indentation. Tabs only

WebViewActivity.java:114, IndentationCheck, Priorität: Normal
'if' child have incorrect indentation level 20, expected level should be one of the following: 5, 6, 7.

WebViewActivity.java:115, RegexpSinglelineJavaCheck, Priorität: Normal
Spaces arent allowed for indentation. Tabs only

WebViewActivity.java:115, IndentationCheck, Priorität: Normal
'if rcurly' have incorrect indentation level 16, expected level should be one of the following: 4, 5, 6.

WebViewActivity.java:117, IndentationCheck, Priorität: Normal
'if' have incorrect indentation level 16, expected level should be one of the following: 4, 5, 6.

WebViewActivity.java:117, RegexpSinglelineJavaCheck, Priorität: Normal
Spaces arent allowed for indentation. Tabs only

WebViewActivity.java:118, RegexpSinglelineJavaCheck, Priorität: Normal
Spaces arent allowed for indentation. Tabs only

WebViewActivity.java:118, IndentationCheck, Priorität: Normal
'if' child have incorrect indentation level 20, expected level should be one of the following: 5, 6, 7.

WebViewActivity.java:119, RegexpSinglelineJavaCheck, Priorität: Normal
Spaces arent allowed for indentation. Tabs only

WebViewActivity.java:119, IndentationCheck, Priorität: Normal
'if' child have incorrect indentation level 20, expected level should be one of the following: 5, 6, 7.

WebViewActivity.java:120, RegexpSinglelineJavaCheck, Priorität: Normal
Spaces arent allowed for indentation. Tabs only

WebViewActivity.java:120, IndentationCheck, Priorität: Normal
'if rcurly' have incorrect indentation level 16, expected level should be one of the following: 4, 5, 6.

WebViewActivity.java:120, RegexpMultilineCheck, Priorität: Normal
Next line is an unnecessary newline

WebViewActivity.java:120, RegexpMultilineCheck, Priorität: Normal
Maximum of blank lines is 1

WebViewActivity.java:209, IndentationCheck, Priorität: Normal
'if rcurly' have incorrect indentation level 12, expected level should be 3.

WebViewActivity.java:209, RegexpSinglelineJavaCheck, Priorität: Normal
Spaces arent allowed for indentation. Tabs only

WebViewActivity.java:214, IndentationCheck, Priorität: Normal
'if' have incorrect indentation level 12, expected level should be 3.

WebViewActivity.java:214, RegexpSinglelineJavaCheck, Priorität: Normal
Spaces arent allowed for indentation. Tabs only

WebViewActivity.java:215, IndentationCheck, Priorität: Normal
'if' child have incorrect indentation level 16, expected level should be 4.

WebViewActivity.java:215, RegexpSinglelineJavaCheck, Priorität: Normal
Spaces arent allowed for indentation. Tabs only

WebViewActivity.java:216, IndentationCheck, Priorität: Normal
'if' child have incorrect indentation level 16, expected level should be 4.

WebViewActivity.java:216, RegexpSinglelineJavaCheck, Priorität: Normal
Spaces arent allowed for indentation. Tabs only

WebViewActivity.java:217, IndentationCheck, Priorität: Normal
'if' child have incorrect indentation level 16, expected level should be 4.

WebViewActivity.java:217, RegexpSinglelineJavaCheck, Priorität: Normal
Spaces arent allowed for indentation. Tabs only

WebViewActivity.java:218, IndentationCheck, Priorität: Normal
'if rcurly' have incorrect indentation level 12, expected level should be 3.

WebViewActivity.java:218, RegexpSinglelineJavaCheck, Priorität: Normal
Spaces arent allowed for indentation. Tabs only
```

**Comment by FleetAdmiral** [comment]
@cheidenreich Ah most of them seem to be indentation faults (Android studio kept recommending me to switch to 4 spaces over tabs, so i changed them all :/). I'll fix them ASAP :)

**Comment by FleetAdmiral** [comment]
@cheidenreich I've fixed the warnings. Could you please look into the same and let me know if any problems persist? 
Also, any way i can run my code locally to get the same warnings so that i can fix them before i submit PRs?

Thanks!

**Comment by robertpainsi** [comment]
@FleetAdmiral, I tested and reviewed your code yesterday but it took me some time to do some additional review. Always keep in mind that I'm mostly giving feedback to code you didn't write.

develop branch:
Currently `WebChromeClient.onProgressChanged` is used to handle the loading dialog. This is not the right method, or at least there is a better place.
When opening a program, `onProgressChanged` is also triggered twice (maybe because of an iframe), showing the loading dialog twice. I suggest we drop the whole `WebChromeClient`.

The [`WebViewClient`](https://developer.android.com/reference/android/webkit/WebViewClient.html) has two better methods, `onPageStarted` and `onPageFinished`.
* Use `onPageStarted` to create and show the loading dialog (please use `loadingDialog` as name).
* Use `onPageFinished` to dismiss the loading dialog.

The behavior will exactly be the same as in the current develop branch except that it fixes the bug described above.

@FleetAdmiral. your branch:
* When tapping on `Explore`, the loading dialog disappears only after the whole page has been completely loaded.
* When tapping on any program (or any other link), the loading dialog doesn't appear at all.

There are a few ways to dismiss the loading dialog earlier. I'll just mention two.

* In [video.mp4](https://jira.catrob.at/secure/attachment/13396/13396_update.v1.mp4), `WebViewClient.onPageCommitVisible` is used. It looks perfectly fine, even with 2g (maybe there is caching involved!). API level 23 isn't a problem at all. However, according to the documentation, the moment the method is called may be just too early.
* Per JS Callback onDocumentReady. However, implementing it this way isn't worth it at all.

So, if there isn't a good and simple solution, we should drop the early dismissing after all. I'm still thinking about what we can and should do here…

> Also, any way i can run my code locally to get the same warnings so that i can fix them before i submit PRs?

The gradle task `check`, e.g. `./gradlew clean check`

**Comment by FleetAdmiral** [comment]
@robertpainsi Hey! I just re-checked my build and found the following:

1) In my branch, the loading dialog disappears early only from the second time that you hit 'explore' without closing the app. Weird enough, after pressing explore once (which loads as before), in subsequent selection of 'Explore', the loading dialog disappears early. I will have to look into why its behaving so. Yes i noticed that the loading dialog doesn't appear when clicking on other links, and i've figured out why that is so (mostly due to how the 'webLoadStatus' variable is behaving). I'll change the same. If i get to figure out why the loading dialog doesn't appear the first time, i don't see why my solution wouldn't work. But if i find another easier solutions (i'm playing around with WebViewClient right now), i guess that'd be better.

2) Regarding WebViewClient.onPageCommitVisible, i don't see why we can't use it because the docs mention 'This method is called when the body of the HTTP response has started loading, is reflected in the DOM...' which means that the function would be called only after we get information about the next page we're navigating to, so it sounds right in my opinion. In fact, i'd rather implement this over my method. Is there something i'm missing here? (I wouldn't want to implement it using JS callback either :P)

3) In the develop branch, as you mentioned, i think we'd be better off loading the dialog within the WebViewClient and be done with WebChromeClient in general. I'm ready to take this up too. Should i make a new issue out of this, or work it into this PR itself?

**Comment by FleetAdmiral** [comment]
@robertpainsi Hey! my mid terms are going on as of now, so i won't be able to work on this till the 19th, hope thats fine :).
Also, could you please look into the queries i asked in points 2 and 3^?

Thanks :)

**Comment by robertpainsi** [comment]
> Also, could you please look into the queries i asked in points 2 and 3^?

On it! I'll reply on it in the next days.

> my mid terms are going on as of now, so i won't be able to work on this till the 19th, hope thats fine :).

Sure, best luck! :smiley_cat:

**Comment by robertpainsi** [comment]
The documentation of `WebViewClient.onPageCommitVisible` also states that 'The next draw will display either the !background color of the WebView!…' This is the only thing I'm concerned about. :worried: 

> done with WebChromeClient in general

What do you think about dropping the `WebChromeClient` entirely by using `WebViewClient.onPageStarted` and `WebViewClient.onPageFinished`?

**Comment by FleetAdmiral** [comment]
@robertpainsi Since it looks like caching is being used to load previously loaded content, or "some of the contents of the newly loaded page", even if it was to load earlier than required, it'd probably either load only the background color or previously cached content. I haven't implemented it personally yet, i'll do so today, but since you seemed to have tried it out and it works, i think it'd be fine to go ahead with it :D.
By "done with WebChromeClient in general", i was suggesting dropping the use of WebChromeClient entirely (seems redundant to use both WebViewClient and WebChromeClient), and push all content into methods of WebViewClient :)
Let me know what you feel about going ahead with both of the above, and i'll work accordingly. Thanks :)

**Comment by robertpainsi** [comment]
k, lets give it a shot! :smile: 

> i was suggesting dropping the use of WebChromeClient entirely 

Yes, great! :+1:

**Comment by FleetAdmiral** [comment]
@robertpainsi  I'm made both the changes (dropping WebChromeClient as well as using `WebViewClient.onPageCommitVisible` . Currently, implementation with onPageCommitVisible works such that it either gives the required result, or gives https://jira.catrob.at/secure/attachment/14748/vid_2.mp4 
As i predicted, in worst case it simply loads the background color of the WebView and waits for page to load (which didn't sound like a problem to me).
Let me know what you think of this implementation, and i'll commit it if you find it ok :)

**Comment by FleetAdmiral** [comment]
@robertpainsi  Any updates on this?

**Comment by robertpainsi** [comment]
> `WebViewClient.onPageCommitVisible`

Thanks for providing those videos. They help a lot.
I would like to move the discussion about `WebViewClient.onPageCommitVisible` back to the Jira ticket because I suspect that `onPageCommitVisible` may dismiss the loading dialog just to early on slow internet or [other reasons](https://bugs.chromium.org/p/chromium/issues/detail?id=458566#c3).

I also would like to merge your other cool stuff already! :smile: 
* Background color
* Non cancelable loading dialog
* > dropping the use of WebChromeClient entirely

Let me know what you want to handle in this PR and when I can start reviewing your code. :mag:

**Comment by FleetAdmiral** [comment]
@robertpainsi I've made a new commit with only the changes requested:

> Background color
Non manually cancelable loading dialog
dropping the use of WebChromeClient entirely

The changes regarding CAT-2298 have been removed. Please do review the code :D


> Thanks for providing those videos. They help a lot.

Yea, i've been looking at how the other members share information and learning to do the same :)


> I would like to move the discussion about WebViewClient.onPageCommitVisible back to the Jira ticket because I suspect that onPageCommitVisible may dismiss the loading dialog just to early on slow internet or other reasons.

The video i uploaded is how it behaves on 2G too, the only thing that happens when it loads too early is the background color is shown to be the same as the WebView color (i've uploaded 2 videos on Jira regarding the same). I'll follow up on the same over there :)

File: catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java [hunk 2]
```diff
@@ -64,6 +64,7 @@
 	public static final String MEDIA_FILE_PATH = "media_file_path";
 	public static final String CALLING_ACTIVITY = "calling_activity";
 	private static final String FILENAME_TAG = "fname=";
+	private ProgressDialog progressCircle;
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java]: Please rename. `loadingDialog`, `webpageLoadingDialog`, `webviewLoadingDialog` or something better you come up with. The name should indicate that it's the loading dialog for webpages.

File: catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java [hunk 3]
```diff
@@ -125,7 +104,7 @@ public void onProgressChanged(WebView view, int progress) {
 		webView.setDownloadListener(new DownloadListener() {
 			@Override
 			public void onDownloadStart(String url, String userAgent, String contentDisposition, String mimetype,
-					long contentLength) {
+										long contentLength) {
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java]: Please do not change indentation level.

File: catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java [hunk 4]
```diff
@@ -189,8 +168,21 @@ public boolean onKeyDown(int keyCode, KeyEvent event) {
...
 		@Override
 		public void onPageStarted(WebView view, String urlClient, Bitmap favicon) {
+			if (progressCircle == null) {
+				progressCircle = new ProgressDialog(view.getContext(), R.style.WebViewLoadingCircle);
+				progressCircle.setCancelable(true);
+				progressCircle.setCanceledOnTouchOutside(false);
+				progressCircle.setProgressStyle(android.R.style.Widget_ProgressBar_Small);
+				try {
+					progressCircle.show();
+				} catch (Exception e) {
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java]: `.show()` doesn't through an exception.

File: catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java (hunk 3 above)
Comment (FleetAdmiral) [inline catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java]: Sorry, had auto-indented after finishing code, will fix :)

File: catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java (hunk 2 above)
Comment (FleetAdmiral) [inline catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java]: Alright, will do!

**Comment by FleetAdmiral** [comment]
@robertpainsi  Since i've already made the PR, is it possible for me to clean up my git commits?
Sorry for the trouble, i had made the PR only after committing what i thought was initially required but then things added on, hence :P
I'll keep it in mind from the next time :)
Working on the issue mentioned, will commit it soon :D

**Comment by robertpainsi** [comment]
> Since i've already made the PR, is it possible for me to clean up my git commits?

Yes, by using interactive rebasing (squashing and reordering) and force pushing. Let me know if you need guidance. :smiley:

File: catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java (hunk 4 above)
Comment (FleetAdmiral) [inline catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java]: Could you elaborate on this please?

Comment (FleetAdmiral) [inline catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java]: Got what you meant, but this code existed from before, should i still change it?

Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java]: This catch just hurts sooo bad every time I see it :weary:

Comment (FleetAdmiral) [inline catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java]: challenge_accepted

Will do! I'm done with all changes, so i'll make a commit for the same, then clean my commits! Sorry for the trouble!

**Comment by FleetAdmiral** [comment]
@robertpainsi I seem to be done! Please do proceed as you planned :)

File: catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java (hunk 3 above)
Comment (FleetAdmiral) [inline catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java]: Fixed :)

**Comment by robertpainsi** [comment]
@FleetAdmiral, I'll clean up your commits (whitespaces, commit message, …), so I can merge it. I just need to know one thing. You changed your name and email address during the PR.

Author: Pranav Nair <FleetAdmiral@users.noreply.github.com>
Commit: FleetAdmiral <pranavnair97@gmail.com>

Which one do you want to use? (I'll remove the other one from the commits and also from this comment)

**Comment by FleetAdmiral** [comment]
@robertpainsi I'd like to use pranavnair97@gmail.com (i think it changed because some commit in between had been made directly on Github's file editing interface).
Thanks!

**Comment by robertpainsi** [comment]
@FleetAdmiral, I didn't close it, GitHub did automatically :rofl:

I was using [this new GitHub feature](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) but pushed the wrong hash (the current develop branch). So there were no commits to merge and GitHub closed the PR. Since it was closed, GitHub also removed the right to push to your branch. So I couldn't change the hash anymore.

Long story short, stupid GitHub (and me, but just a little bit :wink:), I created a new PR https://github.com/Catrobat/Catroid/pull/2216.
@FleetAdmiral, you can find your old changes [here](https://github.com/FleetAdmiral/Catroid/commit/fa31d03) or [here](https://github.com/Catrobat/Catroid/compare/develop...robertpainsi:CAT-2296_improve_web_loading_screen_backup?expand=1) if you still need them. I'll also retest everything quickly today in a few hours.

**Comment by FleetAdmiral** [comment]
@robertpainsi Haha, ok! I'll follow up on the other PR :)

**Comment by robertpainsi** [comment]
@FleetAdmiral, I tested it again and it looks good. :+1: Please take a look at the changes, commits and messages. Also give a thumbs up on the PR if you're fine with the changes.

I also fixed a little bug (`NullPointerException`, added a null check [here](https://github.com/Catrobat/Catroid/pull/2216/files#diff-75d178c7983184b0b7f3a31befe7f051R190), hash changes triggers `onPageFinished` but *not* `onPageStarted`)

**Comment by FleetAdmiral** [comment]
@robertpainsi Yep, those changes look fine to me :D

**Pull request closed without merging**